### PR TITLE
optimize away SortNode in case it is covered by an arangosearch view

### PR DESCRIPTION
**Pull request by jsteemann**
this implementation is a stub with hard-coded attribute names

**WIP - no need to review right now**

### CODE REVIEW

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/4054/

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 1]
```diff
@@ -120,6 +122,117 @@ bool optimizeSearchCondition(IResearchViewNode& viewNode, Query& query, Executio
...
+
+    if (sortCondition.numAttributes() > primarySort.size()) {
+      // the SORT condition in the query has more attributes than the view
+      // is sorted by. we cannot optimize in this case
+      return false;
+    }
+    
+    // check if all sort conditions match
+    for (size_t i = 0; i < sortElements.size(); ++i) {
+      if (sortElements[i].ascending != primarySort.directon(i)) {
```
Comment (jsteemann) [inline arangod/Aql/IResearchViewOptimizerRules.cpp]: should be `direction`
```suggestion
      if (sortElements[i].ascending != primarySort.direction(i)) {
```

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 2]
```diff
@@ -120,6 +122,117 @@ bool optimizeSearchCondition(IResearchViewNode& viewNode, Query& query, Executio
...
+          current->getType() == EN::INDEX ||
+          current->getType() == EN::COLLECT ||
+          current->getType() == EN::SORT) {
+        // any of these node types will lead to more/less results in the output,
+        // and may as well change the sort order, so let's better abort here
+        return false;
+      }
+    }
+  
+    LOG_TOPIC("affe4", WARN, arangodb::Logger::FIXME) << "found a SORT statement (#" << sortNode->id() << ") to optimize away";
```
Comment (jsteemann) [inline arangod/Aql/IResearchViewOptimizerRules.cpp]: this LOG message was there just for testing. we should remove it now the feature is there.
```suggestion
```

File: arangod/IResearch/IResearchViewMeta.h [hunk 3]
```diff
@@ -72,6 +73,83 @@ struct IResearchViewMeta {
...
+      return _fields;
+    }
+
+    std::vector<basics::AttributeName> const& field(size_t i) const noexcept {
+      TRI_ASSERT(i < this->size());
+
+      return _fields[i];
+    }
+
+    bool directon(size_t i) const noexcept {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchViewMeta.h]: should be `direction`
```suggestion
    bool direction(size_t i) const noexcept {
```

**Pull request merged**